### ISSUE TITLE
Initial debug configurations

### ISSUE THREAD

**Issue by bersbersbers** [ManuelHentschel/VSCode-R-Debugger#99]
**Describe the bug**
I had trouble setting up the extension, because 
1. I expected the extension to come with a basic debug configuration, but
2. I did not see it - I had to add one to `launch.json` myself.

I found this, so you are probably aware of automatic debug configurations:
https://github.com/ManuelHentschel/VSCode-R-Debugger/blob/4fd0e703860a53cffbbc9fe2d086bd949098a078/src/extension.ts#L25
(**Update**: I guess I was wrong here: you're not using `DebugConfigurationProviderTriggerKind.Dynamic` nor `onDebugDynamicConfigurations` anywhere.)

But as I said, I don't see them. This is after I click "Show all automatic debug configurations":
![image](https://user-images.githubusercontent.com/12128514/96567513-272d3c80-12c7-11eb-9e73-1c06977684ed.png)

What works is pressing F5 - then, I get this:
![image](https://user-images.githubusercontent.com/12128514/96567630-4cba4600-12c7-11eb-9765-2abfcdb7d6a2.png)

but then this:
![image](https://user-images.githubusercontent.com/12128514/96567687-5d6abc00-12c7-11eb-9524-f44c52978699.png)

and this:
![image](https://user-images.githubusercontent.com/12128514/96567757-75424000-12c7-11eb-8f75-72eabd692862.png)

So then I opened `C:\Temp`. Debug configurations are still empty.

Then, upon F5, I get this:
![image](https://user-images.githubusercontent.com/12128514/96567906-a1f65780-12c7-11eb-97de-6c57c8ac8194.png)
(Which is weird because `C:\Program Files\R\R-4.0.2\bin\R.exe` is exactly the path my R binary.)

Still, clicking "Open launch.json", it opens a `launch.json` with three entries pre-filled (`Launch Workspace`, `Debug R-File`, `Debug R-Function`). And now I also see them in the launch panel, and debugging works without changing any path.

In summary:
- One needs to open a folder workspace for debugging to work.
- Something is weird with the R path before the debug config is added
- Automatic debug configs do not seem to work(?). Compare
https://code.visualstudio.com/updates/v1_45#_automatic-debug-configurations
https://code.visualstudio.com/updates/v1_45#_providing-debug-configurations-dynamically

**Comment by ManuelHentschel** [comment]
Thanks for raising this! I made some changes regarding this topic recently, so please make sure to install from master (the marketplace version is a bit behind currently), which might address some of the problems.

Currently, my "workflow" to add an initial debug configuration is to
1. Open an empty folder workspace
2. Click `create a launch.json file` in the debug menu
3. Select `R Debugger`, which leaves me with the following config:

``` json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "R-Debugger",
            "request": "launch",
            "name": "Launch Workspace",
            "debugMode": "workspace",
            "workingDirectory": "${workspaceFolder}",
            "allowGlobalDebugging": true
        },
        {
            "type": "R-Debugger",
            "request": "launch",
            "name": "Debug R-File",
            "debugMode": "file",
            "workingDirectory": "${workspaceFolder}",
            "file": "${file}",
            "allowGlobalDebugging": true
        },
        {
            "type": "R-Debugger",
            "request": "launch",
            "name": "Debug R-Function",
            "debugMode": "function",
            "workingDirectory": "${fileDirname}",
            "file": "${file}",
            "mainFunction": "main",
            "allowGlobalDebugging": false
        },
        {
            "type": "R-Debugger",
            "request": "attach",
            "name": "Attach to R process",
            "splitOverwrittenOutput": true
        }
    ]
}
```

I agree that this is a bit static and rather unintuitive, if you are used to dynamic debug configuration providers.

I'll look into the articles you linked, and see what I can do :)

**Comment by bersbersbers** [comment]
Oh yes, now that you remind me:
- The link "create a launch.json file" does nothing for me (maybe due to an old extension version)
- If you have any other debug configuration for any other language (Python, in my case), you don't get this:
![image](https://user-images.githubusercontent.com/12128514/96581720-8e082100-12da-11eb-9b4f-2fa324d19ac2.png)
you get this instead:
![image](https://user-images.githubusercontent.com/12128514/96581759-9a8c7980-12da-11eb-8376-c7b0330d4e45.png)

And so there is neither "create a launch.json file" nor "Show all automatic ones".

**Comment by ManuelHentschel** [comment]
>     * The link "create a launch.json file" does nothing for me (maybe due to an old extension version)

Is nothing happening at all, or only after selecting "R Debugger" in the prompt?


>     * If you have any other debug configuration for any other language (Python, in my case), you don't get this:
>      ...
>       you get this instead:
>      ...
> And so there is neither "create a launch.json file" nor "Show all automatic ones".

The linked PR adds configuration snippets, that can be added to an existing launch.json.

**Comment by ManuelHentschel** [comment]
The linked PR adds a dynamic debug configuration provider and configuration snippets to extend an existing launch.json.

I am not aware of other ways to provide debug configurations (other than maybe writing custom scripts, which is probably not recommended).

If you know any or have problems/suggestions regarding the added features, please let me know :)

**Comment by bersbersbers** [comment]
> > ```
> > * The link "create a launch.json file" does nothing for me (maybe due to an old extension version)
> > ```
> 
> Is nothing happening at all, or only after selecting "R Debugger" in the prompt?

Nothing at all. It's a link that can be clicked (as seen by a border appearing), but clicking it does not do anything.

**Comment by bersbersbers** [comment]
> ![image](https://user-images.githubusercontent.com/12128514/96567906-a1f65780-12c7-11eb-97de-6c57c8ac8194.png)
> (Which is weird because `C:\Program Files\R\R-4.0.2\bin\R.exe` is exactly the path my R binary.)

Spaces may be part of the reason - this is what I get upon "Install Debug package":

```
C:\Users\bers>C:\Program Files\R\R-4.0.2\bin\R.exe --no-save -e "install.packages('https://github.com/ManuelHentschel/VSCode-R-Debugger/releases/download/v0.3.1/vscDebugger_0.3.1.zip', 
repos=NULL)" -e "install.packages('jsonlite', repos='http://cran.r-project.org')" -e "install.packages('R6', repos='http://cran.r-project.org')"
'C:\Program' is not recognized as an internal or external command,
operable program or batch file.
```

It may be relevant that I am using `"terminal.integrated.shell.windows": "cmd.exe"`

**Comment by bersbersbers** [comment]
> The linked PR adds a dynamic debug configuration provider and configuration snippets to extend an existing launch.json.

I will certainly try this, thanks! It's still a pain to install an extension from GitHub (I need to clone, build in VS Code, package with npm, and then install, right)?

**Comment by ManuelHentschel** [comment]
> 
> 
> > > ```
> > > * The link "create a launch.json file" does nothing for me (maybe due to an old extension version)
> > > ```
> > 
> > 
> > Is nothing happening at all, or only after selecting "R Debugger" in the prompt?
> 
> Nothing at all. It's a link that can be clicked (as seen by a border appearing), but clicking it does not do anything.

If you don't even get to select "R Debugger" this appears to be a more general issue. Unless of course the issue goes away by disabling the extension, which would be quite odd...

**Comment by ManuelHentschel** [comment]
> 
> 
> > ...
> > (Which is weird because `C:\Program Files\R\R-4.0.2\bin\R.exe` is exactly the path my R binary.)
> 
> Spaces may be part of the reason - this is what I get upon "Install Debug package":
> 
> ```
> C:\Users\bers>C:\Program Files\R\R-4.0.2\bin\R.exe --no-save -e "install.packages('https://github.com/ManuelHentschel/VSCode-R-Debugger/releases/download/v0.3.1/vscDebugger_0.3.1.zip', 
> repos=NULL)" -e "install.packages('jsonlite', repos='http://cran.r-project.org')" -e "install.packages('R6', repos='http://cran.r-project.org')"
> 'C:\Program' is not recognized as an internal or external command,
> operable program or batch file.
> ```
> 
> It may be relevant that I am using `"terminal.integrated.shell.windows": "cmd.exe"`

You need to put quotes around the path, if it contains spaces:
```
"rdebugger.rterm.windows": "\"Path containing spaces\""
```

In most cases, it should also be possible to leave this entry blank and let the extension find the R executable in the registry or PATH environment variable.

You're not the first person with this problem, though, so we could add this info to the setting description. Adding the spaces automatically might work as well, but then it wouldn't be possible to pass e.g. command line arguments as part of the path

**Comment by ManuelHentschel** [comment]
> I will certainly try this, thanks! It's still a pain to install an extension from GitHub (I need to clone, build in VS Code, package with npm, and then install, right)?

We are using Github actions to make this process significantly easier. The clone/build/package part is done after each push and the produced .vsix files can be downloaded as artifact e.g. [here](https://github.com/ManuelHentschel/VSCode-R-Debugger/actions/runs/317622845) for the PR linked above.

Similarly, the R package with compiled binaries (for windows and mac) can be found e.g. [here](https://github.com/ManuelHentschel/vscDebugger/actions/runs/316581039). Though I would probably recommend using devtools + Rtools to compile the package yourself, if you plan on working with development versions.

**Comment by bersbersbers** [comment]
> > > ...
> > > (Which is weird because `C:\Program Files\R\R-4.0.2\bin\R.exe` is exactly the path my R binary.)
> > 
> > 
> > Spaces may be part of the reason - this is what I get upon "Install Debug package":
> > ```
> > C:\Users\bers>C:\Program Files\R\R-4.0.2\bin\R.exe --no-save -e "install.packages('https://github.com/ManuelHentschel/VSCode-R-Debugger/releases/download/v0.3.1/vscDebugger_0.3.1.zip', 
> > repos=NULL)" -e "install.packages('jsonlite', repos='http://cran.r-project.org')" -e "install.packages('R6', repos='http://cran.r-project.org')"
> > 'C:\Program' is not recognized as an internal or external command,
> > operable program or batch file.
> > ```
> > 
> > 
> > It may be relevant that I am using `"terminal.integrated.shell.windows": "cmd.exe"`
> 
> You need to put quotes around the path, if it contains spaces:
> 
> ```
> "rdebugger.rterm.windows": "\"Path containing spaces\""
> ```

I forgot to say that I don't seem to be setting this variable anywhere. `R.exe` is in my `%PATH%`, so it's probably found there. We might branch this off into a separate issue ;)

**Comment by bersbersbers** [comment]
> > I will certainly try this, thanks! It's still a pain to install an extension from GitHub (I need to clone, build in VS Code, package with npm, and then install, right)?
> 
> We are using Github actions to make this process significantly easier. The clone/build/package part is done after each push and the produced .vsix files can be downloaded as artifact e.g. [here](https://github.com/ManuelHentschel/VSCode-R-Debugger/actions/runs/318294618) for the PR linked above.

Ah, great. It's not too bad after all: clone the repository, open in VS Code, and use the "Run Extension" debug configuration. That  part works fine.

I see that the extensions starts with data from the `test` subfolder, which has a `launch.json`. I think the goal of this issue is to have debug configurations without this, so I deleted it. 

1. Then, when running the extension, I was a bit disappointed because I still only get this:
![image](https://user-images.githubusercontent.com/12128514/96684713-5487f200-137c-11eb-8833-be3f2b6f99e2.png)

I had expected automatic debug configurations to be listed immediately, but this does not seem to be the case when you don't have any other configuration, even after opening a `.R` file. This experience is still suboptimal, but that's how it works in VS Code apparently.

At least, this time, "Show all automatic debug configurations" shows
![image](https://user-images.githubusercontent.com/12128514/96684814-7e411900-137c-11eb-917a-b7c06c863562.png)
and then
![image](https://user-images.githubusercontent.com/12128514/96684987-c7916880-137c-11eb-9261-39c29ac7616f.png)
This is neat, because it allows to run (and add to your launch.json) only single configurations, and not the whole bunch.

2. By contrast, when you have any debug config to start with, you get this, beautifully:
![image](https://user-images.githubusercontent.com/12128514/96685199-13dca880-137d-11eb-875f-54788c322c48.png)

So I guess this works as expected. Thank you!

**Comment by ManuelHentschel** [comment]
> I forgot to say that I don't seem to be setting this variable anywhere. `R.exe` is in my `%PATH%`, so it's probably found there. We might branch this off into a separate issue ;)

I just added automatic quotes if the path contains spaces and an option to add custom command line arguments for the call to R. Thah should hopefully work for everyone.

**Comment by ManuelHentschel** [comment]
>     1. Then, when running the extension, I was a bit disappointed because I still only get this:

I think I finally figured out, how to make the "Run and Debug" button work with R (see latest push to the PR). Now the button launches the debugger in "Debug file" mode, if you have an R file open.

**Comment by ManuelHentschel** [comment]
> I had expected automatic debug configurations to be listed immediately, but this does not seem to be the case when you don't have any other configuration, even after opening a `.R` file. This experience is still suboptimal, but that's how it works in VS Code apparently.

This is probably intentional, since listing all automatic debug configurations would require activating all installed debugger extensions, which might take some time

**Comment by ManuelHentschel** [comment]
Thanks again for opening this issue and providing such detailed feedback! :) 
This topic is quite important for new users and I probably wouldn't have found the different ways to provide debug configs myself, since I have a launch.json in all my projects anyways.

**Comment by bersbersbers** [comment]
> > ```
> > 1. Then, when running the extension, I was a bit disappointed because I still only get this:
> > ```
> 
> I think I finally figured out how to make the "Run and Debug" button work with R (see latest push to the PR). Now the button launches the debugger in "Debug file" mode, if you have an R file open.

Interesting! You mean https://github.com/ManuelHentschel/VSCode-R-Debugger/pull/100/commits/4d90f0283939bc2b34d51cdecb7259264e5b0ce2? 

I cannot really see what that has changed (maybe because I did not try the version just before). I noticed that my VS Code installation in "Extension Host" mode behaves weirdly (remember https://github.com/ManuelHentschel/VSCode-R-Debugger/issues/99#issuecomment-712828287), and there are other issues sometimes (such as a Run button being visible, not no dropdown - just one example). But my normal installation is working now.The whole VS Code debugging configuration experience seems to be pretty convoluted, depending on which extensions are activated, which files are and have been open, whether one is in file, folder or (multi-root) workspace mode, etc. etc. (hence my interest in https://github.com/microsoft/vscode/issues/109083, by the way).

> Thanks again for opening this issue and providing such detailed feedback! :)

You're very welcome - it's a very helpful extension to me, so helping out is worth my time :)

> This topic is quite important for new users and I probably wouldn't have found the different ways to provide debug configs myself, since I have a launch.json in all my projects anyways.

I agree. Maybe it makes sense to test on a fresh VS Code from time to time, in an environment that has never seen the extension and does not have things like `launch.json` - or a `settings.json`.

Using the latest version, I receive repeated notifications about changing my shell environment (related to `VSCODE_R_DEBUGGER_TERMINAL_ID`) which I would like to ignore, but they are pretty persistent. You seem to have this setting in your `settings.json`, so maybe you don't even get it anymore.

**Comment by ManuelHentschel** [comment]
> Interesting! You mean [4d90f02](https://github.com/ManuelHentschel/VSCode-R-Debugger/commit/4d90f0283939bc2b34d51cdecb7259264e5b0ce2)?

Yep

> I cannot really see what that has changed (maybe because I did not try the version just before). I noticed that my VS Code installation in "Extension Host" mode behaves weirdly (remember [#99 (comment)](https://github.com/ManuelHentschel/VSCode-R-Debugger/issues/99#issuecomment-712828287)), and there are other issues sometimes (such as a Run button being visible, not no dropdown - just one example). But my normal installation is working now.The whole VS Code debugging configuration experience seems to be pretty convoluted, depending on which extensions are activated, which files are and have been open, whether one is in file, folder or (multi-root) workspace mode, etc. etc. (hence my interest in [microsoft/vscode#109083](https://github.com/microsoft/vscode/issues/109083), by the way).

As far as I'm aware, the changes to package.json only take effect when installing the extension, not when launched in a debug session. To make sure you see all changes, you can use the [default build task](https://github.com/ManuelHentschel/VSCode-R-Debugger/blob/1d804ed46ffe2bca21f0a8f544f719be6d27af20/.vscode/tasks.json#L17-L23) to build and install the extension before debugging it.

**Comment by ManuelHentschel** [comment]
> Using the latest version, I receive repeated notifications about changing my shell environment (related to `VSCODE_R_DEBUGGER_TERMINAL_ID`) which I would like to ignore, but they are pretty persistent. You seem to have this setting in your `settings.json`, so maybe you don't even get it anymore.

That setting is related to flow control (next/continue/step) when debugging in attached mode. The setting is changed each time a terminal is created to be able to match the R process running in the terminal with the terminal itself. But I think I'll disable this feature by default, since the notification can be quite annoying...

**Comment by bersbersbers** [comment]
> I just added automatic quotes if the path contains spaces and an option to add custom command line arguments for the call to R. Thah should hopefully work for everyone.

My issue may be different overall - same error message with quotes:
![image](https://user-images.githubusercontent.com/12128514/96753863-c5f19000-13d0-11eb-844d-1cb304017a1d.png)

Is maybe the 2000ms timeout too short?

**Comment by ManuelHentschel** [comment]
You could try setting a longer timeout, e.g. `"rdebugger.timeouts.startup": 9000`.
Or add the command line argument `"rdebugger.rterm.args": ["--no-init-file"]`, if you have stuff in your .Rprofile that might take some time.
